ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,545 @@
+  -> LMDB.Transaction mode (Maybe DbState)
+readDbSettingsMaybeNull db = LMDB.get db ()
+
+withDbSettingsRW ::
+     LMDB.Database () DbState
+  -> (DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState))
+  -> LMDB.Transaction LMDB.ReadWrite a
+withDbSettingsRW db f = withDbSettingsRWMaybeNull db $ maybe (Exn.throw DbErrNoSettings) f
+
+withDbSettingsRWMaybeNull ::
+      LMDB.Database () DbState
+   -> (Maybe DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState))
+   -> LMDB.Transaction LMDB.ReadWrite a
+withDbSettingsRWMaybeNull db f  =
+  readDbSettingsMaybeNull db >>= f >>= \(r, new_s) -> LMDB.put db () (Just new_s) $> r
+
+data GuardDbDir  = GDDMustExist | GDDMustNotExist
+
+guardDbDir :: MonadIO m => GuardDbDir -> FS.SomeHasFS m -> FS.FsPath -> m FilePath
+guardDbDir gdd (FS.SomeHasFS fs) path = do
+  FS.doesFileExist fs path >>= \b -> when b $ Exn.throw $ DbErrStr $ "guardDbDir:must be a directory:" <> show path
+  FS.doesDirectoryExist fs path >>= \case
+    True | GDDMustNotExist <- gdd ->
+           -- TODO Should throw a DbErrDbExists exception
+           -- Callers should delete the directory if they want to restore a snapshot
+           FS.removeDirectoryRecursive fs path
+    False | GDDMustExist <- gdd -> Exn.throw $ DbErrDbDoesntExist path
+    _ -> pure ()
+  FS.createDirectoryIfMissing fs True path
+  FS.unsafeToFilePath fs path
+
+{-------------------------------------------------------------------------------
+ Initialize an LMDB
+-------------------------------------------------------------------------------}
+
+-- | How to initialise an LMDB Backing store
+data LMDBInit l
+  = LIInitialiseFromMemory (WithOrigin SlotNo) (LedgerTables l ValuesMK) -- ^ Initialise with these values
+  | LIInitialiseFromLMDB FS.FsPath -- ^ Initialise by copying from an LMDB db at this path
+  | LINoInitialise -- ^ The database is already initialised
+
+-- | Initialize the LMDB from these provided values
+initFromVals ::
+     TableStuff l
+  => WithOrigin SlotNo
+  -> LedgerTables l ValuesMK
+  -> Db l
+  -> IO ()
+initFromVals dbsSeq vals Db{..} = LMDB.readWriteTransaction dbEnv $
+  withDbSettingsRWMaybeNull dbSettings $ \case
+    Nothing -> zipLedgerTablesA initLMDBTable dbBackingTables vals $> ((), DbState{dbsSeq})
+    Just _ -> Exn.throw $ DbErrStr "initFromVals: db already had state"
+
+initFromLMDBs :: MonadIO m
+  => Trace.Tracer IO TraceDb
+  -> FS.SomeHasFS m
+  -> FS.FsPath
+  -> FS.FsPath
+  -> m ()
+initFromLMDBs tracer shfs from0 to0 = do
+  liftIO $ Trace.traceWith tracer $ TDBInitialisingFromLMDB from0
+  from <- guardDbDir GDDMustExist shfs from0
+  to <- guardDbDir GDDMustNotExist shfs to0
+  liftIO $ Exn.bracket
+    (LMDB.openEnvironment from defaultLMDBLimits) -- TODO assess limits, in particular this could fail if the db is too big
+    LMDB.closeEnvironment $ \e -> lmdbCopy tracer e to
+
+
+lmdbCopy :: MonadIO m
+  => Trace.Tracer m TraceDb
+  -> LMDB.Environment LMDB.ReadWrite
+  -> FilePath
+  -> m ()
```

Comment:
Also I'm unsure if when we use this, we are actually storing the new environment somewhere or is it being lost? I imagined `lmdbCopy` should return `m (LMDB.Environment LMDB.ReadWrite)` so that we keep track of the new copied DB, and in case we are not interested, we close that environment afterwards. Or maybe not?

Reply:
Might be. Returning the environment will require using the resource registry right? Since in this case we won't be able to use `bracket`.